Docs: how to run services reliably and update service autorestart to service lifecycle. (#541)
canonical/pebble: docs/reference/service-lifecycle.md

# Service lifecycle

> Pebble manages the lifecycle of a service, including starting, stopping, and restarting it, with a focus on handling health checks and failures, and implementing auto-restart with backoff strategies, which are achieved using a state machine with the following states:

Reviewer: This sentence is quite long. Would it be possible to split it? Something like this:
```suggestion
Pebble manages the lifecycle of a service, including starting, stopping, and restarting it. Pebble also handles health checks, failures, and auto-restart with backoff. This is all achieved using a state machine with the following states:
```

- starting: The service is in the process of starting.
- running: The `okayDelay` (see below) period has passed, and the service runs normally.
- terminating: The service is being gracefully terminated.
- killing: The service is being forcibly killed.
- stopped: The service has stopped.
- backoff: The service will be put in the backoff state before the next start attempt if the service is configured to restart when it exits.
- exited: The service has exited (and won't be automatically restarted).

## Service start

A service begins in an "initial" state. Pebble tries to start the service's underlying process and transitions the service to the "starting" state.

## Start confirmation

Pebble waits for a short period (`okayDelay`, defaults to one second) after starting the service. If the service runs without exiting after the `okayDelay` period, it's considered successfully started, and the service's state is transitioned into "running".

No matter if the service is in the "starting" or "running" state, if you get the service, the status will be shown as "active". Read more in the [`pebble services`](#reference_pebble_services_command) command.

## Start failure

> If the service exits quickly, the started channel receives an error. The error, along with the last logs, are added to the task (see more in [Changes and tasks](/reference/changes-and-tasks.md)). This also ensures logs are accessible.

Reviewer: What does "started channel" refer to?

Author: This is a bit too deep into the code, I will rephrase this.

## Abort start

If the user interrupts the start process (e.g., with a SIGKILL), the service transitions to stopped, and a SIGKILL signal is sent to the underlying process.

## Auto-restart

> By default, Pebble's service manager automatically restarts services that exit unexpectedly, regardless of whether the service is in the "starting" state (the `okayDelay` period has not passed) or in the "running" state (`okayDelay` is passed, and the service is considered to be "running").
> 
> This is done whether the exit code is zero or non-zero, but you can fine-tune the behaviour using the `on-success` and `on-failure` fields in a configuration layer. The possible values for these fields are:

Reviewer: The first paragraph is a nice summary of the behaviour - I think it will stick in my head :smile: 

I have a small suggestion for the second paragraph, as "This is done whether the exit code is zero or non-zero" could be misread as meaning something happens regardless of whether the exit code is zero or non-zero.

```suggestion
By default, Pebble's service manager automatically restarts services that exit unexpectedly, regardless of whether the service is in the "starting" state (the `okayDelay` period has not passed) or in the "running" state (`okayDelay` has passed, and the service is considered to be "running").

Pebble considers a service to have exited unexpectedly if the exit code is non-zero. You can fine-tune the behaviour using the `on-success` and `on-failure` fields in a configuration layer. The possible values for these fields are:
```

* `restart`: restart the service and enter a restart-backoff loop (the default behaviour).
* `shutdown`: shut down and exit the Pebble daemon (with exit code 0 if the service exits successfully, exit code 10 otherwise)
  - `success-shutdown`: shut down with exit code 0 (valid only for `on-failure`)
  - `failure-shutdown`: shut down with exit code 10 (valid only for `on-success`)
* `ignore`: ignore the service exiting and do nothing further

## Backoff

Pebble implements a backoff mechanism that increases the delay before restarting the service after each failed attempt. This prevents a failing service from consuming excessive resources.

The `backoff-delay` defaults to half a second, the `backoff-factor` defaults to 2.0 (doubling), and the increasing delay is capped at `backoff-limit`, which defaults to 30 seconds. All of the three configurations can be customized, read more in [Layer specification](../reference/layer-specification).

> For example, with default settings for the above configuration, in `restart` mode, the first time a service exits, Pebble waits for half a second. If the service exits again, Pebble calculates the next backoff delay by multiplying the current delay by `backoff-factor`, which results in a 1-second delay. The next delay will be 2 seconds, then 4 seconds, and so on, capped at 30 seconds.

Reviewer: Very minor - we can drop "for example" to help with making the sentence shorter. (This part is a description of the default behaviour, so I think it's fine to not say "for example")
```suggestion
With default settings for the above configuration, in `restart` mode, the first time a service exits, Pebble waits for half a second. If the service exits again, Pebble calculates the next backoff delay by multiplying the current delay by `backoff-factor`, which results in a 1-second delay. The next delay will be 2 seconds, then 4 seconds, and so on, capped at 30 seconds.
```
